source/common/config/grpc_stream.h

```diff
@@ -131,11 +141,74 @@ class GrpcStream : public Grpc::AsyncStreamCallbacks<ResponseProto>,
     return false;
   }
 
+  absl::optional<Grpc::Status::GrpcStatus> getCloseStatus() { return close_status_; }
+
 private:
   void setRetryTimer() {
     retry_timer_->enableTimer(std::chrono::milliseconds(backoff_strategy_->nextBackOffMs()));
   }
 
+  // https://github.com/envoyproxy/envoy/issues/14591
+  // Log level should be reduced when the remote close failure is `Ok` or is retriable and has only
+  // been occurring for a short amount of time.
+  void logClose(Grpc::Status::GrpcStatus status, const std::string& message) {
+    if (Grpc::Status::WellKnownGrpcStatus::Ok == status) {
+      ENVOY_LOG(debug, CloseLogMessage, service_method_.name(), status, message);
+      return;
+    }
+
+    if (!onlyWarnOnRepeatedFailure(status)) {
+      // Failure is considered non-retriable. Warn.
+      ENVOY_LOG(warn, CloseLogMessage, service_method_.name(), status, message);
+      return;
+    }
+
+    if (!isCloseStatusSet()) {
+      // First failure. Debug. Record occurrence.
+      ENVOY_LOG(debug, CloseLogMessage, service_method_.name(), status, message);
+      setCloseStatus(status, message);
+      return;
+    }
+
+    uint64_t ms_since_first_close = std::chrono::duration_cast<std::chrono::milliseconds>(
+                                        time_source_.monotonicTime() - close_time_)
+                                        .count();
+    Grpc::Status::GrpcStatus close_status = close_status_.value();
+
+    if (status != close_status) {
+      // This is a different failure. Warn on both statuses and remember the new one.
+      ENVOY_LOG(warn, CloseLogMessageWithPrevious, service_method_.name(), status, message,
+                close_status, close_message_, ms_since_first_close);
+      setCloseStatus(status, message);
+      return;
+    }
+
+    if (ms_since_first_close > RetryMaxDelayMs) {
+      // Warn if we are over the time limit.
+      ENVOY_LOG(warn, CloseLogMessageWithSince, service_method_.name(), ms_since_first_close,
+                close_status, close_message_);
+      return;
+    }
+
+    // Failure is retriable and new enough to only log at the debug level.
+    ENVOY_LOG(debug, CloseLogMessage, service_method_.name(), status, message);
+  }
+
+  bool onlyWarnOnRepeatedFailure(Grpc::Status::GrpcStatus status) {
+    return Grpc::Status::WellKnownGrpcStatus::Unavailable == status ||
+           Grpc::Status::WellKnownGrpcStatus::DeadlineExceeded == status ||
+           Grpc::Status::WellKnownGrpcStatus::Internal == status;
```

Comment:
This is included due to the example in #14591 (error 13), but [gRPC docs](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) classify this as a serious error:
>Internal errors. This means that some invariants expected by the underlying system have been broken. This error code is reserved for serious errors.

I'm questioning whether the server should return this in the first place if it doesn't seem to be serious. @howardjohn @kyessenov @mandarjog thoughts?

Reply:
I am pretty sure `StreamAggregatedResources gRPC config stream closed: 13,` comes from keepalive closing the connection in the bug. I am a bit surprised it shows up as `13`, and not `Unavailable`. It may be worth exploring a bit more.

FWIW if you want to test you can set `--keepaliveMaxServerConnectionAge=5s` on Istiod to get an XDS server that closes connection this way. If you use out-of-the-box Istio we have an XDS proxy the translates the error to `OK` anyways though

Reply:
Thank you. That makes sense to me since it's happening every 30 minutes in that example. It looks like in recent Istio (1.10.2) the recurring error is now 14 (Unavailable)
>2021-07-12T23:58:48.196332Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, transport is closing

After building an Envoy with this change and not including this line/treating Internal as retriable, I no longer got warnings for the above error. So I think we should not special case Internal for now and only DeadlineExceeded/Unavailable